src/application/predictor.hpp

```diff
@@ -104,12 +108,45 @@ class Predictor {
       Log::Fatal("Could not recognize the data format of data file %s", data_filename);
     }
 
+    std::unordered_map<int, int> feature_names_map_;
+    if(has_header) {
+      std::ifstream tmp_file;
+      tmp_file.open(data_filename);
+      if (!tmp_file.is_open()) {
+        Log::Fatal("Data file %s doesn't exist'", data_filename);
+      }
+      std::string line1;
+      if (!tmp_file.eof()) {
+        std::getline(tmp_file, line1);
+        std::vector<std::string> header = Common::Split(line1.c_str(), "\t");
+        header.erase(header.begin() + boosting_->LabelIdx());
+        for(int i = 0; i < static_cast<int>(header.size()); ++i) {
+          for(int j = 0; j < static_cast<int>(boosting_->FeatureNames().size()); ++j) {
+            if(header[i] == boosting_->FeatureNames()[j]) {
+              feature_names_map_[i] = j;
+              break;
+            }
+          }
+        }
+      }
+    }
     // function for parse data
     std::function<void(const char*, std::vector<std::pair<int, double>>*)> parser_fun;
     double tmp_label;
-    parser_fun = [this, &parser, &tmp_label]
+    parser_fun = [this, &parser, &tmp_label, &has_header, &feature_names_map_]
     (const char* buffer, std::vector<std::pair<int, double>>* feature) {
       parser->ParseOneLine(buffer, feature, &tmp_label);
+      if(has_header) {
+        for(int i = 0; i < static_cast<int>(feature->size()); ++i) {
+          if(feature_names_map_.find((*feature)[i].first) != feature_names_map_.end()) {
+            (*feature)[i].first = feature_names_map_[(*feature)[i].first];
+          }
+          else {
+            (*feature).erase((*feature).begin() + i);
```

Comment:
erase in the loop ?

Reply:
yes, here is for deleting the features which are not used in the model but exist in the test data.

Reply:
I don't think it is safe to do this. 
A better solution is allocating another map that used to store non-deleted items.

Reply:
I have moved the non-used features to the end of the feature vector, I think this way is also a safe solution.